.github/scripts/transform_dbt_test_results.py

```diff
@@ -0,0 +1,1082 @@
+
+    def to_dict(self) -> typing.Dict:
+        """Serialize the TestCategory object as a dictionary."""
+        return {
+            "category": self.category,
+            "test_results": [result.to_dict() for result in self.test_results],
+        }
+
+    def __repr__(self) -> str:
+        num_failing_rows = sum(
+            len(result.failing_rows) for result in self.test_results
+        )
+        return (
+            f"TestCategory(category={self.category!r}, "
+            f"status={self.status!r}, "
+            f"num_tests={len(self.test_results)}, "
+            f"num_failing_rows={num_failing_rows})"
+        )
+
+    @classmethod
+    def from_dict(cls, category_dict: typing.Dict) -> "TestCategory":
+        """Deserialize a TestCategory object from a dictionary."""
+        return TestCategory(
+            category=category_dict["category"],
+            results=[
+                TestResult.from_dict(result_dict)
+                for result_dict in category_dict["test_results"]
+            ],
+        )
+
+    @property
+    def fieldnames(self) -> typing.List[str]:
+        """Get a list of fieldnames that encapsulates all of the fieldnames
+        for all of the rows of tests tracked by this group."""
+        fieldnames = []
+        for result in self.test_results:
+            for fieldname in result.fieldnames:
+                if fieldname not in fieldnames:
+                    fieldnames.append(fieldname)
+
+        # Remove any fixed fieldnames from the ordered list that are not
+        # present in this group
+        fixed_field_order = [
+            field for field in self._fixed_fieldnames if field in fieldnames
+        ]
+
+        # Reorder the fieldnames so that diagnostic fields are presented in the
+        # correct order
+        for field in reversed(fixed_field_order):
+            fieldnames.insert(0, fieldnames.pop(fieldnames.index(field)))
+
+        return fieldnames
+
+    @property
+    def rows(self) -> typing.List[typing.List]:
+        """Format the rows of tests tracked by this group, with
+        fieldname data excluded. The combination of this property and the
+        `fieldnames` property can be used to write to a csv.Writer or
+        to an openpyxl.Workbook sheet for the tests tracked by this group."""
+        fieldnames = self.fieldnames
+        return [
+            [row.get(fieldname) for fieldname in fieldnames]
+            for result in self.test_results
+            for row in result.failing_rows
+        ]
+
+    @property
+    def status(self) -> Status:
+        """Return an aggregate status for this category based on the statuses
+        of its TestResult objects."""
+        return (
+            Status.PASS
+            if all(
+                result.status == Status.PASS for result in self.test_results
+            )
+            else Status.FAIL
+        )
+
+    def add_to_workbook(self, workbook: openpyxl.Workbook) -> None:
```

Comment:
This function is adapted from the previous version of the `format_dbt_test_failures` script, where it was called [`add_sheet_to_workbook`](https://github.com/ccao-data/data-architecture/blob/66c85bfa973d163bd29c4e41f08ecf58eafff08d/.github/scripts/format_dbt_test_failures.py#L564). I've moved it to a method on the `TestCategory` class since it is essentially a function to add a `TestCategory` object to an existing workbook.